sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EventHandlerDelegateShouldHaveProperArguments.cs

```diff
@@ -0,0 +1,128 @@
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class EventHandlerDelegateShouldHaveProperArguments : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4220";
+        private const string MessageFormat = "{0}";
+        private const string NullEventArgsMessage = "Use 'EventArgs.Empty' instead of null as the event args of " +
+            "this event invocation.";
+        private const string NullSenderMessage = "Make the sender on this event invocation not null.";
+        private const string NonNullSenderMessage = "Make the sender on this static event invocation null.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> EventHandlerTypes = new HashSet<KnownType>
+        {
+            KnownType.System_EventHandler,
+            KnownType.System_EventHandler_TEventArgs
+        };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var invocation = (InvocationExpressionSyntax)c.Node;
+                    if (invocation.ArgumentList == null ||
+                        invocation.ArgumentList.Arguments.Count != 2)
+                    {
+                        return;
+                    }
+
+                    var methodSymbol = c.SemanticModel.GetSymbolInfo(invocation).Symbol as IMethodSymbol;
+                    if (methodSymbol == null ||
+                        methodSymbol.MethodKind != MethodKind.DelegateInvoke ||
+                        !methodSymbol.ContainingType.ConstructedFrom.IsAny(EventHandlerTypes))
+                    {
+                        return;
+                    }
+
+                    var isSecondParameterNull = invocation.ArgumentList.Arguments[1].Expression
+                        .RemoveParentheses()
+                        .IsKind(SyntaxKind.NullLiteralExpression);
+                    if (isSecondParameterNull)
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, invocation.GetLocation(), NullEventArgsMessage));
+                    }
+
+                    var eventSymbol = GetEventSymbol(invocation.Expression, c.SemanticModel);
+                    if (eventSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var isFirstParameterNull = invocation.ArgumentList.Arguments[0].Expression
```

Comment:
Could that method be generic? If so, the rule could also check for that.

If this can be generic, we could also look for default(T) in here as well.